src/rgw/store/dbstore/config/sqlite.cc

```diff
@@ -0,0 +1,2052 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2022 Red Hat, Inc.
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation. See file COPYING.
+ *
+ */
+
+#include <charconv>
+#include <initializer_list>
+#include <map>
+
+#undef FMT_HEADER_ONLY
+#define FMT_HEADER_ONLY 1
+#include <fmt/format.h>
+
+#include <sqlite3.h>
+
+#include "include/buffer.h"
+#include "include/encoding.h"
+#include "common/dout.h"
+#include "common/random_string.h"
+#include "rgw_zone.h"
+
+#include "common/connection_pool.h"
+#include "sqlite/connection.h"
+#include "sqlite/error.h"
+#include "sqlite/statement.h"
+#include "sqlite_schema.h"
+#include "sqlite.h"
+
+#define dout_subsys ceph_subsys_rgw_dbstore
+
+namespace rgw::dbstore::config {
+
+struct Prefix : DoutPrefixPipe {
+  std::string_view prefix;
+  Prefix(const DoutPrefixProvider& dpp, std::string_view prefix)
+      : DoutPrefixPipe(dpp), prefix(prefix) {}
+  unsigned get_subsys() const override { return dout_subsys; }
+  void add_prefix(std::ostream& out) const override {
+    out << prefix;
+  }
+};
+
+namespace {
+
+// parameter names for prepared statement bindings
+static constexpr const char* P1 = ":1";
+static constexpr const char* P2 = ":2";
+static constexpr const char* P3 = ":3";
+static constexpr const char* P4 = ":4";
+static constexpr const char* P5 = ":5";
+static constexpr const char* P6 = ":6";
+
+
+void read_text_rows(const DoutPrefixProvider* dpp,
+                    const sqlite::stmt_execution& stmt,
+                    std::span<std::string> entries,
+                    sal::ListResult<std::string>& result)
+{
+  result.entries = sqlite::read_text_rows(dpp, stmt, entries);
+  if (result.entries.size() < entries.size()) { // end of listing
+    result.next.clear();
+  } else {
+    result.next = result.entries.back();
+  }
+}
+
+struct RealmRow {
+  RGWRealm info;
+  int ver;
+  std::string tag;
+};
+
+void read_realm_row(const sqlite::stmt_execution& stmt, RealmRow& row)
+{
+  row.info.id = sqlite::column_text(stmt, 0);
```

Comment:
Using readable strings instead of direct numbers while making prepared statements or reading the row values, helped me to better understand & debug the issues (esp., when the table schema changes)

Reply:
yeah, name-based column lookups is what i started with. but when i was putting together the postgres backend, i noticed this warning in the [pqxx docs for row::operator[](zview col_name)](https://libpqxx.readthedocs.io/en/stable/a01148.html#aee26781d8c0000bdc1d80c1624b17c81):
> This is much slower than indexing by number, or iterating.

but that's unlikely to matter with these small tables, so i can change it back 👍

Reply:
> yeah, name-based column lookups is what i started with. but when i was putting together the postgres backend, i noticed this warning in the [pqxx docs for row::operator[](zview col_name)](https://libpqxx.readthedocs.io/en/stable/a01148.html#aee26781d8c0000bdc1d80c1624b17c81):

Interesting! SQLite allows column lookup based on only the index number. (https://www.sqlite.org/c3ref/column_blob.html)
I had used enum ( https://github.com/soumyakoduri/ceph/blob/dbstore_object/src/rgw/store/dbstore/sqlite/sqliteDB.cc#L152) to give names to these indexes and avoid any mistakes in the ordering while reading values.

As you mentioned, since these are small tables, I guess its okay to read directly based on numbers.

Reply:
> I had used enum ( https://github.com/soumyakoduri/ceph/blob/dbstore_object/src/rgw/store/dbstore/sqlite/sqliteDB.cc#L152) to give names to these indexes and avoid any mistakes in the ordering while reading values.

ohh, i see. my first draft had used `sqlite3_column_name()` on each column and did string comparisons based on that. it seems like the enum approach would have the same issues around schema changes - you'd still have to keep the enums in-sync with the queries. i think it's just as easy to update the column indices here

Reply:
Thats right! Just that it was easy to read from enum and map them (https://github.com/soumyakoduri/ceph/blob/dbstore_object/src/rgw/store/dbstore/sqlite/sqliteDB.cc#L281) esp., when any new column was added to the middle of the table.

But yes..as you mentioned, for these smaller tables, its easy to update indices.